File: Data Structures & Algorithms/house-robber/submission-1.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        cache = {}

        i = 2
        cache[0] = nums[0]

        if len(nums) == 1:
            return cache[0]

        cache[1] = max(nums[0], nums[1]) 
        
        if len(nums) == 2:
            return cache[1]

        while i  < len(nums):
            cache[i] = max(nums[i] + cache[i - 2], cache[i-1]) 
            i += 1 

        return cache[len(nums) - 1]

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-1.cpp
class Solution {
   public:
    vector<int> productExceptSelf(vector<int>& nums) {
        // initialize an array of all 1s
        vector<int> result(nums.size(), 1);

        int leftProduct = 1;
        int rightProduct = 1;
        // forward pass, calculate the left product and multiply the current value
        for (int i = 0; i < nums.size(); i++) {
            result[i] *= leftProduct;
            leftProduct *= nums[i];
        }

        // reverse pass, calculate the left product and multiply the current value
        for (int i = nums.size() - 1; i >= 0; i--) {
            result[i] *= rightProduct;
            rightProduct *= nums[i];
        }

        return result;
    }
};
